buildspec.yml

```diff
@@ -1,25 +1,26 @@
 version: 0.2
-
 phases:
   install:
     runtime-versions:
       nodejs: 16
       python: 3.10
     commands:
-      - python3 -m pip install -U pip setuptools virtualenv
-      - npm install --location=global npm@latest
+    - python3 -m pip install -U pip setuptools virtualenv
+    - npm install --location=global npm@latest
   build:
     commands:
-      - cd ./deployment
-      - ./build-s3-dist.sh -b $DIST_OUTPUT_BUCKET -v $VERSION
-      - ./build-open-source-dist.sh $SOLUTION_NAME
-      - ./run-unit-tests.sh
+    - cd source
+    - npm ci
+    - cd ../deployment
+    - ./build-s3-dist.sh -b $DIST_OUTPUT_BUCKET -v $VERSION
+    - ./build-open-source-dist.sh $SOLUTION_NAME
+    - ./run-unit-tests.sh
```

Comment:
Why does adding npm ci here specifically improve the build process?

should this not be the responsibility of the build-s3-dist.sh file so that the file doesn't rely on this command as a hidden pre-requisite?

also you may be able to remove the build-open-source-dist script entirely. I was able to for Instance Scheduler at least

Reply:
Adding `npm ci` here does not improve the pipeline build (except for the additional safety of `ci` over `i`), it maintains the same behavior. The motivation for moving this outside of `build-s3-dist.sh` is to make the required loop for build->test->deploy as short as possible. I know locally I have to install first, and that's not exactly a surprise, there are other "hidden" pre-requisites such as having node and npm installed. But taking `ci` out of the build script saves me 15 seconds on every build if I've already run it. That's a big deal for my ability to iterate quickly. If this solution ever gets to a point where it can be built and deployed outside of this script, it will become irrelevant.

I know we can make open-source a no-op, but that change doesn't really provide any value so it's not a priority.

Reply:
hmm, I'm not sure yet whether I agree with the extra 15 seconds is worth the extra dependency on the environment, but it's not exactly a deal-breaker either. Thanks for explaining.